Fix CUDA/HIP math function annotations and cover it by unit tests

sbastrakov opened the pull request:
First commit is the tests, second is potential fix by @SimeonEhrig @psychocoderHPC .

~~The tests are done in a bit of quick-and-dirty fashion, wanted to make a quick PR to pass it through CI.~~ Ready for  review

TODO:

- [x] Beautify the tests while CI is running
- [x] I messed up authorship of the second commit, will fix.

j-stephan commented on include/alpaka/math/MathUniformCudaHipBuiltIn.hpp:
```diff
@@ -18,6 +18,7 @@
 #include <alpaka/math/Complex.hpp>
 #include <alpaka/math/Traits.hpp>
 
+#include <algorithm>
```
Why is this required?

sbastrakov replied:
Oh idk. @SimeonEhrig do you?

SimeonEhrig replied:
That is a mistake. I forgot to remove it.

sbastrakov committed:
Extend math unit tests to also check a case of wrapping lambda

The base testing template got extended with a new parameter to control it.
All types always test the non-wrapped (as before this commit) case.
Conditionally enable the wrapped case when alpaka should support it.

SimeonEhrig committed:
Add back __host__ __device__ annotations to CUDA/HIP math implementations

Fixes #1688

sbastrakov commented:
I now put this lambda to existing base math unit test fixture, so that it always applies to all tests for math functions that also checks the value. An unfortunate side product of that change is that on the dev system now it compiles even without the fix! In that case, it fails runtime with `invalid device function`. I guess if we want a hard compilation error, I can also add a single test like in my first attempt, but for a fixed math function. That would be a relatively small test then (also in terms of future maintenance).

sbastrakov commented:
Oh no, some CI jobs failed. I will have a look tomorrow morning.

sbastrakov commented:
The current CI issue for clang-cuda seems to be caused by complex number math on CPU. I am now working on that, hopefully fix it soon by adding similar extra function annotations.

(The previously applied fix solved it for real numbers and tests now cover that).

sbastrakov committed:
Apply a similar fix for alpaka::Complex<> native math implementations

j-stephan reviewed (approved):
LGTM.

sbastrakov commented:
Unfortunately, yet another fix is required as one more thing appeared. Working on it. I've also realized that this PR was opened to 0.9.0-rc1 branch, and not to `dev`. So should I close this PR and open a new one to `develop` with the fixed version?

j-stephan commented:
No, the release is more important right now. We'll cherry-pick afterwards.

sbastrakov committed:
Fix float math tests accidentally using double

sbastrakov commented:
A new attempt is up. Along the way I also found another issue with `float` (real-number) tests using `double`. I accidentally made this mistake when introducing tests for complex numbers (that required to change a lot of generic math unit tests things). I made the fix as a separate commit in this PR.

sbastrakov committed:
Slightly extend accepted margin for math unit tests

sbastrakov commented:
@j-stephan ready for re-review. CI finally passed.
